Added ocspsign API endpoint. Support for ocspResponse in cfssljson

fastest963 opened the pull request:
This is only my second PR so I'm not very familiar with the codebase as a whole but this PR does:
* Added `ocspsign` API endpoint to sign ocsp responses
* Added support for `ocspResponse` in cfssljson
* cfssljson can also now output to stdout using `-stdout`
* `ocspserve` will not silently ignore trailing whitespace (or double spaces) in the responses file
* Updated README with example for creating a OCSP responses file from `ocspsign` cli.

I don't know if this is okay but I'm getting an error from openssl when parsing the OCSP responses from `ocspsign`:
```
# cfssl ocspsign -ca ca.crt -responder resp.crt -responder-key resp.key -cert test.crt | cfssljson -bare ocsp
# openssl ocsp -text -respin ocsp-response.der
OCSP Response Data:
    OCSP Response Status: successful (0x0)
Error parsing response
140127808935840:error:0D0680A8:asn1 encoding routines:ASN1_CHECK_TLEN:wrong tag:tasn_dec.c:1342:
140127808935840:error:0D06C03A:asn1 encoding routines:ASN1_D2I_EX_PRIMITIVE:nested asn1 error:tasn_dec.c:854:
140127808935840:error:0D08303A:asn1 encoding routines:ASN1_TEMPLATE_NOEXP_D2I:nested asn1 error:tasn_dec.c:774:Field=producedAt, Type=OCSP_RESPDATA
140127808935840:error:0D08303A:asn1 encoding routines:ASN1_TEMPLATE_NOEXP_D2I:nested asn1 error:tasn_dec.c:774:Field=tbsResponseData, Type=OCSP_BASICRESP
140127808935840:error:0D0C706E:asn1 encoding routines:ASN1_item_unpack:decode error:asn_pack.c:189:
    Response Type: Basic OCSP Response
```
Should I file that as a new bug or is that expected?

fastest963 commented on api/ocsp/ocspsign.go:
```diff
@@ -0,0 +1,96 @@
+		return err
+	}
+
+	pemResponse := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: resp})
```
I'm not sure if that's the right block type (it doesn't seem like it to me) but I just copied that from cli.go.

lziest replied:
I agree, this is incorrect. I think base64 encoding the response is good enough.

fastest963 replied:
Should I change cli.go too?

lziest commented on README.md:
````diff
@@ -230,7 +230,7 @@ certificate.
 
 ```
-cfssl gencert -remote=remote_server [-hostname=comma,separated,hostnames] csrjson
+cfssl gencert -remote=remote_server [-hostname=comma,separated,hostnames] | csrjson
````
This is incorrect, but I admit it is a little confusing. It would be better to put it "csr.json"

fastest963 replied:
Woops yes this is confusing I'll change it to csr.json

lziest commented on README.md:
````diff
@@ -239,12 +239,24 @@ a certificate. You may use `-hostname` to override certificate SANs.
 
 ```
-cfssl gencert -ca cert -ca-key key [-hostname=comma,separated,hostnames] csrjson
+cfssl gencert -ca cert -ca-key key [-hostname=comma,separated,hostnames] | csrjson
````
ditto

lziest commented on cmd/cfssljson/cfssljson.go:
```diff
@@ -39,9 +41,17 @@ type Response struct {
 func main() {
 	bare := flag.Bool("bare", false, "the response from CFSSL is not wrapped in the API standard response")
 	inFile := flag.String("f", "-", "JSON input")
+	output := flag.Bool("out", false, "output the response instead of saving to a file")
```
do you like "stdout" instead of "out"?

fastest963 replied:
It doesn't matter to me.

lziest commented on cli/ocspsign/ocspsign.go:
```diff
@@ -71,5 +71,16 @@ func ocspSignerMain(args []string, c cli.Config) (err error) {
+func SignerFromConfig(c cli.Config) (ocsp.Signer, error) {
+	//if this is called from serve then we need to use the specific responder key file
+	//fallback to key for backwards-compatibility
+	k := c.ResponderKeyFile
```
IMHO, the fallback is confusing. If we want to use responder-key, let's make it across the board.

fastest963 replied:
Ya I don't disagree but I didnt want to break backwards compatibility. Can we put a depreciation warning and a todo to remove it?

lziest commented on README.md:
````diff
@@ -239,12 +239,24 @@ a certificate. You may use `-hostname` to override certificate SANs.
+
+```
+cfssl ocspsign -ca cert -responder key -responder-key key -cert cert \
+ | csrjson -bare -out -base64 >> responses
````
csrjson -> cfssljson

fastest963 commented:
Updated the PR to address @lziest's suggestions. I also fixed the README changes and the linter complaints. The response is now base64 from `ocspsign` from the cli and the api.

lziest commented:
Thanks, I will test it through and get back to you as soon as possible.

lziest commented on cli/cli.go:
```diff
@@ -174,11 +174,11 @@ func PrintCert(key, csrBytes, cert []byte) {
+// ocspResponse is base64 encoded
 func PrintOCSPResponse(resp []byte) {
-	pemResponse := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: resp})
+	b64Resp := base64.StdEncoding.EncodeToString(resp)
```
LGTM, @jsha do you have any idea that this change may affect your code base?

jsha replied:
Thanks for the ping! I believe this won't affect the Boulder code base, since we call cfssl directly as a library rather than through the API, for both cert signing and OCSP signing. In case you're curious, here's the relevant section of code where we instantiate both signers: https://github.com/letsencrypt/boulder/blob/master/ca/certificate-authority.go#L126.

lziest commented:
LGTM.

kisom commented:
LGTM.

fastest963 commented:
Thanks @lziest and @kisom for the review and merge. My original comment also brought up openssl failing to verify signed OCSP responses:
```
# cfssl ocspsign -ca ca.crt -responder resp.crt -responder-key resp.key -cert test.crt | cfssljson -bare ocsp
# openssl ocsp -text -respin ocsp-response.der
OCSP Response Data:
    OCSP Response Status: successful (0x0)
Error parsing response
140127808935840:error:0D0680A8:asn1 encoding routines:ASN1_CHECK_TLEN:wrong tag:tasn_dec.c:1342:
140127808935840:error:0D06C03A:asn1 encoding routines:ASN1_D2I_EX_PRIMITIVE:nested asn1 error:tasn_dec.c:854:
140127808935840:error:0D08303A:asn1 encoding routines:ASN1_TEMPLATE_NOEXP_D2I:nested asn1 error:tasn_dec.c:774:Field=producedAt, Type=OCSP_RESPDATA
140127808935840:error:0D08303A:asn1 encoding routines:ASN1_TEMPLATE_NOEXP_D2I:nested asn1 error:tasn_dec.c:774:Field=tbsResponseData, Type=OCSP_BASICRESP
140127808935840:error:0D0C706E:asn1 encoding routines:ASN1_item_unpack:decode error:asn_pack.c:189:
    Response Type: Basic OCSP Response
```

Should I file that as a separate bug or is that intended? (Note: it was doing that before and after my changes)
